win32: Get the string description of the usb serial port via DeviceIoControl.

chinaheyu opened the pull request:
This pull request helps to get the serial number, product string and manufacturer string of the usb serial port correctly on windows.

https://lihashgnis.blogspot.com/2018/07/getting-descriptors-from-usb-device.html?m=1
https://stackoverflow.com/questions/28007468/how-do-i-obtain-usb-device-descriptor-given-a-device-path
https://stackoverflow.com/questions/50701189/how-to-get-usb-device-descriptor-by-doing-deviceiocontrol-directly-on-the-devi

Because the com port description provided by the windows registry is not the usb description we often need, I refer to the above to get the device description of the usb serial port via DeviceIoControl.

#637

chinaheyu commented:
> I also tested it on Windows 10 and got the same result. I'm attaching my usbview report from Windows 11, in case you'd like to investigate, why I didn't get the expected result.
> [USBViewAll.txt](https://github.com/pyserial/pyserial/files/14089859/USBViewAll.txt)
> 

By the way, are you using the latest commit (`c090f45667c68b7a7259a38de7a59b49b016f851`)?

aivarannamaa commented:
> By the way, are you using the latest commit (c090f45667c68b7a7259a38de7a59b49b016f851)?

Yes, I was using https://github.com/pyserial/pyserial/blob/c090f45667c68b7a7259a38de7a59b49b016f851/serial/tools/list_ports_windows.py and https://github.com/pyserial/pyserial/blob/c090f45667c68b7a7259a38de7a59b49b016f851/serial/win32.py for patching my Pyserial.

I'll see later, whether the patched version works with my other devices.

chinaheyu commented:
> > By the way, are you using the latest commit (c090f45667c68b7a7259a38de7a59b49b016f851)?
> 
> Yes, I was using https://github.com/pyserial/pyserial/blob/c090f45667c68b7a7259a38de7a59b49b016f851/serial/tools/list_ports_windows.py and https://github.com/pyserial/pyserial/blob/c090f45667c68b7a7259a38de7a59b49b016f851/serial/win32.py for patching my Pyserial.
> 
> I'll see later, whether the patched version works with my other devices.

I found a similar issue here: https://stackoverflow.com/questions/51025878/how-to-get-string-descriptor-from-a-usb-device-which-is-in-low-power-state

Maybe it's because the device is in a low power state, so it can't communicate properly through IOCTL to get the usb descriptor (as you can see from the usbview output).

I'll try to reproduce and fix this issue later, thank you very much!

aivarannamaa commented:
I now tested your patch with more devices (SAMD-s from Adafruit, Rasperry Pi Pico, several ESP32-S2 and S3 devices, some plain ESP32 devices with CP and CH UART adapters, an ESP8266 with CP2104 UART adapter, a NRF52 device) and found out that with most of these it worked nicely.

The few, that showed proper Product (and Manufacturer) in usbview.exe, but not with your patch, were:

* Lilygo/TTGO T-Micro 32 ESP32 device (expected product value: "CP2104 USB to UART Bridge Controller", actual: None)
* Lolin C3 mini ESP32-C3 (expected product value: "USB JTAG/serial debug unit", actual: None). Curiously, your patch does enhance the value of the Manufacturer field, by changing it from "Microsoft" to "Espressif".
* M5Stack AtomU and AtomLite, ESP32 devices with FT232 or FT234 serial converter (expected value "M5Stack")
* Sparkfun ESP32 Thing (FT X-Series serial converter)
* BBC micro:bit v2. Besides not showing proper Product and Manufacturer fields, your patch turned serial_number field from "99063602000528203977597D73765584000000006E052820" to "" (empty string)

chinaheyu commented:
> I now tested your patch with more devices (SAMD-s from Adafruit, Rasperry Pi Pico, several ESP32-S2 and S3 devices, some plain ESP32 devices with CP and CH UART adapters, an ESP8266 with CP2104 UART adapter, a NRF52 device) and found out that with most of these it worked nicely.
> 
> The few, that showed proper Product (and Manufacturer) in usbview.exe, but not with your patch, were:
> 
> * Lilygo/TTGO T-Micro 32 ESP32 device (expected product value: "CP2104 USB to UART Bridge Controller", actual: None)
> * Lolin C3 mini ESP32-C3 (expected product value: "USB JTAG/serial debug unit", actual: None). Curiously, your patch does enhance the value of the Manufacturer field, by changing it from "Microsoft" to "Espressif".
> * M5Stack AtomU and AtomLite, ESP32 devices with FT232 or FT234 serial converter (expected value "M5Stack")
> * Sparkfun ESP32 Thing (FT X-Series serial converter)
> * BBC micro:bit v2. Besides not showing proper Product and Manufacturer fields, your patch turned serial_number field from "99063602000528203977597D73765584000000006E052820" to "" (empty string)
> 
> Please let me know, if you need usbview report for any of these!

I read the USBView report you provided for the BBC micro:bit v2. It looks like my patch misidentifies the composite device as a noncomposite device. This will result in an error similar to the one below.

Reads the string description as a composite device:

```
device: COM4
name: COM4
hwid: USB VID:PID=303A:4001 SER=123456 LOCATION=1-4:1.0
vid: 303a
pid: 4001
serial_number: 123456
location: 1-4:1.0
manufacturer: Espressif Systems
product: Espressif Device
```

Reads the string description as a noncomposite device:

```
device: COM4
name: COM4
hwid: USB VID:PID=303A:4001 SER=123456 LOCATION=1-4:x.0
vid: 303a
pid: 4001
serial_number: 
location: 1-4:x.0
manufacturer: Microsoft
```

Could you provide me with reports from the [USB Device Tree Viewer](https://www.uwe-sieber.de/usbtreeview_e.html) for these devices? It could provide me with more information to solve the problem. Thanks a lot.

aivarannamaa commented:
Here come the reports from USB Device Tree Viewer:

* [M5Stack Atom U.txt](https://github.com/pyserial/pyserial/files/14112481/M5Stack.Atom.U.txt)
* [M5Stack AtomLite.txt](https://github.com/pyserial/pyserial/files/14112482/M5Stack.AtomLite.txt)
* [Sparfun Thing 32.txt](https://github.com/pyserial/pyserial/files/14112483/Sparfun.Thing.32.txt)
* [BBC MicroBit.txt](https://github.com/pyserial/pyserial/files/14112484/BBC.MicroBit.txt)
* [Lilygo_TTGO T-Micro 32 ESP32.txt](https://github.com/pyserial/pyserial/files/14112485/Lilygo_TTGO.T-Micro.32.ESP32.txt)
* [Lolin C3 mini ESP32-C3.txt](https://github.com/pyserial/pyserial/files/14112486/Lolin.C3.mini.ESP32-C3.txt)

aivarannamaa commented:
Thanks a lot! After your last commit, pyserial reports proper Product, Manufacturer and Serial number for my BBC micro:bit. I'll test with my other boards tomorrow.

Great job!

aivarannamaa commented:
With all my boards I now get proper values for Manufacurer, Product and Serial number. For getting proper Interface, I'm combining your solution with #571.

Thank you!

chinaheyu commented:
> With all my boards I now get proper values for Manufacurer, Product and Serial number. For getting proper Interface, I'm combining your solution with #571.
> 
> Thank you!

I refactored the code and added comments to enhance readability. Thank you for your testing, I think this patch is ready to be merged now.

chinaheyu commented:
> With all my boards I now get proper values for Manufacurer, Product and Serial number. For getting proper Interface, I'm combining your solution with #571.
> 
> Thank you!

I noticed that you want to merge with #571. I've added the feature to read the interface (via `CM_Get_DevNode_Property` instead of `SetupDiGetDeviceProperty`) in the latest commit. So merging #571 may cause a conflict.

I put the old code for getting usb info into [`get_usb_info_from_device_property()`](https://github.com/chinaheyu/pyserial/blob/master/serial/tools/list_ports_windows.py#L680) as a fallback option. I suggest you put the change #571 at the end of the [`get_usb_info_from_device_property()`](https://github.com/chinaheyu/pyserial/blob/master/serial/tools/list_ports_windows.py#L680) function. This avoids conflicts caused by modifying the `iterate_comports()` function at the same time.

aivarannamaa commented:
> I've added the feature to read the interface

That's great! Somehow I missed it. I tested it and it works beautifully. I don't need #571 anymore.

chinaheyu commented:
It is now possible to get the correct interface field on most devices.

waterfallwhitebread commented:
Unfortunately does not seem to work well on devices with multiple serial interfaces, like FT4232H that has 4 UARTs.

Unpatched result:
```
device: COM12
name: COM12
description: USB Serial Port (COM12)
hwid: USB VID:PID=0403:6011 SER=6
vid: 1027
pid: 24593
serial_number: 6
location: None
manufacturer: FTDI
product: None
interface: None
```

Patched result:
```device: COM12
name: USB Serial Port (COM12)
description: Quad RS232-HS - Quad RS232-HS
hwid: USB VID:PID=0403:6011 LOCATION=1-1:1.0
vid: 1027
pid: 24593
serial_number: None
location: 1-1:1.0
manufacturer: FTDI
product: Quad RS232-HS
interface: Quad RS232-HS
```

Location is the same for all COM Ports, e.g.
```
device: COM11
name: USB Serial Port (COM11)
description: Quad RS232-HS - Quad RS232-HS
hwid: USB VID:PID=0403:6011 LOCATION=1-1:1.0
vid: 1027
pid: 24593
serial_number: None
location: 1-1:1.0
manufacturer: FTDI
product: Quad RS232-HS
interface: Quad RS232-HS
```

So there is no way to distinguish between the COM ports by location. For linux this seems to have been solved in PR https://github.com/pyserial/pyserial/pull/141

Here's an example of szHardwareID_str: FTDIBUS\VID_0403+PID_6011+6&1AD5555C&0&1&1\0000

chinaheyu commented:
> Unfortunately does not seem to work well on devices with multiple serial interfaces, like FT4232H that has 4 UARTs.
> 
> Unpatched result:
> 
> ```
> device: COM12
> name: COM12
> description: USB Serial Port (COM12)
> hwid: USB VID:PID=0403:6011 SER=6
> vid: 1027
> pid: 24593
> serial_number: 6
> location: None
> manufacturer: FTDI
> product: None
> interface: None
> ```
> 
> Patched result:
> 
> ```
> name: USB Serial Port (COM12)
> description: Quad RS232-HS - Quad RS232-HS
> hwid: USB VID:PID=0403:6011 LOCATION=1-1:1.0
> vid: 1027
> pid: 24593
> serial_number: None
> location: 1-1:1.0
> manufacturer: FTDI
> product: Quad RS232-HS
> interface: Quad RS232-HS
> ```
> 
> Location is the same for all COM Ports, e.g.
> 
> ```
> device: COM11
> name: USB Serial Port (COM11)
> description: Quad RS232-HS - Quad RS232-HS
> hwid: USB VID:PID=0403:6011 LOCATION=1-1:1.0
> vid: 1027
> pid: 24593
> serial_number: None
> location: 1-1:1.0
> manufacturer: FTDI
> product: Quad RS232-HS
> interface: Quad RS232-HS
> ```
> 
> So there is no way to distinguish between the COM ports by location. For linux this seems to have been solved in PR #141
> 
> Here's an example of szHardwareID_str: FTDIBUS\VID_0403+PID_6011+6&1AD5555C&0&1&1\0000

Since the FTDI driver hides the location information, we had to do some dirty work to get the interface number. I referenced [libusbp](https://github.com/pololu/libusbp/blob/master/src/windows/interface_windows.c#L222) to parse the hardware id to get the interface number. But apparently it doesn't work.

Could you provide me with reports from the [USB Device Tree Viewer](https://www.uwe-sieber.de/usbtreeview_e.html) for your devices?

waterfallwhitebread commented:
Hey @chinaheyu, thanks a lot for reply and the reference to libusbp. Sure, here the reports from USB Device Tree Viewer (just the first 2 UARTs, as the others are accordingly):
![image](https://github.com/pyserial/pyserial/assets/113577187/afa3367a-71f9-4ee2-912b-20fa0ce9158a)
![image](https://github.com/pyserial/pyserial/assets/113577187/ebc93e3e-8bb9-4509-8a6d-342c8df8378a)
![image](https://github.com/pyserial/pyserial/assets/113577187/1dd341bb-12a6-4267-86ab-886bc5f03d5d)
![image](https://github.com/pyserial/pyserial/assets/113577187/0063ec6d-6810-48d8-9295-33c237efec1a)

aivarannamaa commented:
After the last change most of my devices seem to work same as before, but I get following exception with 3 of my ESP32-C3 devices:

```
  File "C:\python_stuff\thonny\thonny\vendored_libs\serial\tools\list_ports_windows.py", line 1182, in comports
    return list(iterate_comports())
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\python_stuff\thonny\thonny\vendored_libs\serial\tools\list_ports_windows.py", line 1156, in iterate_comports
    usb_info = port_device.get_usb_info(device_registry)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\python_stuff\thonny\thonny\vendored_libs\serial\tools\list_ports_windows.py", line 698, in get_usb_info
    language_id = hub_io.suggest_language_id(usb_hub_port)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\python_stuff\thonny\thonny\vendored_libs\serial\tools\list_ports_windows.py", line 870, in suggest_language_id
    return available_languages[0]
           ~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

Here is the UsbTreeview report for one of them: [SeeedXiaoESP32C3.txt](https://github.com/pyserial/pyserial/files/14736239/SeeedXiaoESP32C3.txt)

The weird thing is that I get this exception only sometimes, usually the first time I call `comports()` after plugging in the device -- the next time I call `comports()` it usually works without problems (but not always).

aivarannamaa commented:
@chinaheyu, after latest changes I don't get these errors anymore, but I now discovered that with any of these ESP32-C3 devices plugged in, I now _occasionally_ get long delays before the ports info gets returned.

It looks like there are two methods, which may cause the delay:

* `USBHubDeviceIOControl.request_supported_languages`
* `USBHubDeviceIOControl.request_usb_string_description`

The first call of either method gets stuck in `DeviceIoControl` for 24-30 seconds. The next calls return quickly. For example, if I skip calling `DeviceIoControl` in `request_supported_languages`, the first `request_usb_string_description` becomes slow, etc.

The problem is present also in your older commits, e.g. https://github.com/aivarannamaa/pyserial/commit/bdab39f9f1989daa2fce57b68926806d00b6d0bc#diff-f9966444f1fafc17f351672a463bdcff7833ad6c14a68d9432329a458d1608b0

The next call to `comports()` is quick, unless in the meanwhile I've connected to or disconnected from a serial port in another process.

Unfortunately I can't create a simple script for reproducing this. The earlier I call comports in my app, the less likely the delay becomes.

Can you recommend some tweaks I could try to learn more about this problem?

aivarannamaa commented:
Adding to my last comment -- `DeviceIoControl` call in `request_usb_connection_info` never becomes slow, even though it gets called before `request_supported_languages` and `request_usb_string_description`.

chinaheyu commented:
@aivarannamaa Thank you very much for your feedback.

This is a very peculiar issue that I haven't encountered before. Overall, the problem should be related to the device (ESP32-32 C3) as it seems not to respond with its string description promptly. Subsequent requests are fast because DeviceIoControl caches the result.

After receiving your feedback, I purchased a Seeed Xiao ESP32 C3 board, which should arrive today. I will do my best to resolve this peculiar issue.

Additionally, I used the same API as in USBView, so did you encounter this issue when using USBView?

chinaheyu commented:
@waterfallwhitebread Thank you for providing the screenshot. The latest patch should be able to correctly retrieve the location field of the FT4232H chip.

chinaheyu commented:
@aivarannamaa You are right, if another process opens and closes the serial port, it blocks DeviceIOControl. This phenomenon only occurs on ESP32 C3 devices.

I also found an issue on StackOverflow that may be relevant: https://stackoverflow.com/questions/6749681/freeze-on-serialport-open-deviceiocontrol-getcommstate-with-usbser-sys

I can't fix this, but we can cache the usb info to avoid calling DeviceIOControl every time. The latest commit should be helpful for your problem.

aivarannamaa commented:
@chinaheyu, thanks! Curiously, now I can't reproduce this problem anymore.

PapaNoyel commented:
Any chance it gets merged into pyserial ?
This PR solved a problem I had on windows to differentiate a Nordic PPK2 from a Nordic Dongle as they appeared the same with vanilla pyserial.

Thanks for this chinaheyu !

jepler commented:
Hi! Is there anything that Adafruit can do to move inclusion of this patch forward? It would be very helpful to us.

keirf commented:
I have a report that this method doesn't work with Windows 7 (https://github.com/keirf/greaseweazle/issues/472). Is that unexpected?

chinaheyu commented:
It should theoretically support Windows 7 (though I haven't tested it), as Cfgmgr32 API has been available since Windows 2000. I plan to test it on Windows 7 next to see if I can reproduce the issue. Thank you @keirf .

chinaheyu commented:
> I have a report that this method doesn't work with Windows 7 ([keirf/greaseweazle#472](https://github.com/keirf/greaseweazle/issues/472)). Is that unexpected?

![Screenshot from 2024-08-21 20-27-10](https://github.com/user-attachments/assets/037f7a40-68cc-4ece-93af-43523662f7a3)

Sorry, I can't reproduce the problem. The patch works fine in Windows 7. Can you provide more detailed python version and system information?

The system image used above: Windows 7 Ultimate with Service Pack 1 (7601.24214) (x64)

chinaheyu commented:
Had some difficulty installing python 3.8, so used python 3.6, but the result should be the same.

![Screenshot from 2024-08-21 21-25-21](https://github.com/user-attachments/assets/0e91bf2a-16ec-4088-a51b-9d9ce3005b7a)

32-bit Windows 7 and 32-bit Python

keirf commented:
Okay, I went to the hassle of installing Win7 in a VM to look at this, and what a pain that was! But I *have* reproduced the issue. I think it must depend on the device or its driver. Specifically, my "Greaseweazle" device enumerates as a generic CDC ACM device, and on Windows 7/8 requires use of Zadig to associate the driver usbser.sys (Zadig calls this "USB Serial (CDC)").

Curiously, the default pyserial list_port_windows.py finds the correct product string for this device on Windows 7 ("Greaseweazle"), which it fails to do on Windows 10/11. And on Win7 your list_port_windows.py fails to list it at all. Perhaps it gets filtered out somewhere?

keirf commented:
I also confirmed this issue on Windows 8.1. This is perhaps more related to Zadig, or pre-Win10 usbser.sys.

chinaheyu commented:
@keirf Bug maybe fixed. Could you try the latest commit  (91177bd)?

keirf commented:
> @keirf Bug maybe fixed. Could you try the latest commit ([91177bd](https://github.com/pyserial/pyserial/commit/91177bd485ab595aee83fb372b89fb2da477dee6))?

It works great! Thanks for the very quick fix. I have already ported it into Greaseweazle tools and I will have the original reporter also confirm the fix.

Thanks again!

adamcbraun commented:
Is there any update on this? This functionality (corrected device naming on windows) is actually addressed in a couple of the pending PRs, but does not seem to ever get completed. It would be great if one of these could reach completion as this functionality would be very helpful for a number of different of implementations.

keirf commented:
I have a report of failure with empty port name (https://github.com/keirf/greaseweazle/issues/549) with the following fix: https://github.com/keirf/greaseweazle/pull/550

Could you please consider this fix, or equivalent, for this upstream PR?

chinaheyu commented:
> I have a report of failure with empty port name ([keirf/greaseweazle#549](https://github.com/keirf/greaseweazle/issues/549)) with the following fix: [keirf/greaseweazle#550](https://github.com/keirf/greaseweazle/pull/550)
> 
> Could you please consider this fix, or equivalent, for this upstream PR?

It seems that some serial devices are not assigned a COMx-style port name. I'm not sure what's causing this, but simply skipping these devices can prevent the program from crashing. In theory, even without a port name, we can still open these serial devices via their interface path, which might serve as an alternative when the port name is None (possibly a better solution than just skipping them). That said, keirf/greaseweazle#550 is correct and necessary. Many thanks for the fix.

lambda-aw commented:
Hello,
Is there any progress on the issue?
The branch solves my issue with CP210x on Windows 11 and product field, so interested in any plan here.

Thanks.

kurtmckee commented:
Hi @lukehugh, thanks for the ping. I'm also not a pyserial maintainer.

I've been holding off on reviewing this PR for only one reason: I'm waiting for #838 to merge. #838 introduces a command to generate a "test asset", which is a JSON file that aims to contain all of the available information for a given device (or devices, if needed) on a given system. That's useful both for debugging detection issues (like what @alonbl reported with trailing null characters) as well as for mocking system calls to confirm that pyserial is handling the devices properly.

#838 introduces such files for sysfs-based operating systems (like Linux), but I've intended to expand the command to support Windows once the PR merges, so that I could ask @alonbl and yourself to please generate the test assets for the devices you're working with. With those test asset files in-hand, as well as files from my own devices on Windows, I've intended to build out the Windows side of the test suite.

The Windows code in pyserial is completely untested, and your PR is a significant contribution that inherits that complete lack of testing. I've therefore been waiting to dive into a review of the PR until I can build out some level of device mocking and testing to better understand the before-and-after behavior of your PR.

I hope this explanation is helpful for understanding why I haven't reviewed the PR yet.

alonbl commented:
Hello @kurtmckee ,

There is always a question of CI and hardware tests....

Do we gain value from simulating hardware for synthetic tests or we invest large amount of resources and eventually we do not test the real-world scenarios.

In case of sysfs, it is quite simple as the sysfs abstraction is plain files, so you could commit the relevant sysfs as-is, and only have a `SYSFS_ROOT` environment variable (for example) to scan the custom directory even without monkey patching the code... So you can have one or two tests for sysfs to make sure we read it correctly... this is not all-on-one-multi-platform agenda, just a simple unit test for sysfs.

In case of windows, it is more difficult, as Windows APIs are much more complex. However, these are static, and decent manual tests of the low-level parts should be sufficient.

For sure, having CI or not should not block us from reviewing code...

Regards,
Alon
